jbpm-audit/src/main/java/org/jbpm/process/audit/query/AbstractAuditDeleteBuilderImpl.java

```diff
@@ -32,10 +32,12 @@
 
 public abstract class AbstractAuditDeleteBuilderImpl<T> extends AbstractDeleteBuilderImpl<T> implements AuditDeleteBuilder<T> {
 
-    protected static String ONLY_COMPLETED_PROCESS_INSTANCES = 
+    protected static String ONLY_COMPLETED_PROCESS_INSTANCES_OR_ORPHANS =
             " l.processInstanceId in (select spl.processInstanceId \n"
             + "FROM ProcessInstanceLog spl \n"
-            + "WHERE spl.status in (2, 3))";
+            + "WHERE spl.status in (2, 3)) \n"
+            + "OR l.processInstanceId not in (select spl.processInstanceId \n "
+            + "FROM ProcessInstanceLog spl)";
```

Comment:
I am afraid this can be really heavy operation when there are millions of records in the ProcessInstanceLog table... and since this can be rare case when people clean up process instances without cleaning tasks this can be simply an edge case which might not be worth to impact all typical use cases, wdyt?

Reply:
Do you think it would impact performance a lot when there is already an index on processInstanceId? Right now DB has to check at first if there is a PID in table with only active PI and secondly if it is not present at all in PI log. If there is an index on processInstanceId, it might be quite quick to determine if there is a hit or not. But maybe I look at it very simply. Anyway, at least I tried, but if you think it is not worth it, you can close it, I have no problem with it ;)

Reply:
OK, I did a little research and yes, it is killing performance quite badly - but bear in mind that if there are thousand of running processes, it would kill it too. And what about consistency? Shouldn't AuditTaskImpl processInstanceId reference processInstanceId in log by foreign key? Because if you skip Task log in that command, task would point at non-existent process.

Reply:
OK, apparently this was a very good exercise. NOT IN has to check also NULL values, therefore it has to go through whole table (because 5 != NULL means unknown, which means false in where clause :) ). So I changed it for NOT EXISTS which is much much quicker. After executing scripts for ProcessInstanceLog table which build all indexes, I was able to compare these 2 queries:

```

EXPLAIN DELETE
FROM AuditTaskImpl l
WHERE l.processInstanceId IN (SELECT spl.processInstanceId
                              FROM ProcessInstanceLog spl
                              WHERE spl.status IN ( 2, 3))
      OR
      l.processInstanceId NOT IN (SELECT spl.processInstanceId
                                  FROM ProcessInstanceLog spl);
```
produces
```
Delete on audittaskimpl l  (cost=8.61..60070689.06 rows=38144 width=6)
  ->  Seq Scan on audittaskimpl l  (cost=8.61..60070689.06 rows=38144 width=6)
        Filter: ((hashed SubPlan 1) OR (NOT (SubPlan 2)))
        SubPlan 1
          ->  Index Scan using idx_pinstlog_status on processinstancelog spl  (cost=0.29..8.61 rows=1 width=8)
                Index Cond: (status = ANY ('{2,3}'::integer[]))
        SubPlan 2
          ->  Materialize  (cost=0.00..2234.99 rows=50866 width=8)
                ->  Seq Scan on processinstancelog spl_1  (cost=0.00..1781.66 rows=50866 width=8)

```

Now the correct version using NOT EXISTS instead of NOT IN

```
EXPLAIN DELETE
FROM AuditTaskImpl l
WHERE l.processInstanceId IN (SELECT spl.processInstanceId
                              FROM ProcessInstanceLog spl
                              WHERE spl.status IN (2,3))
      OR NOT EXISTS (SELECT spl.processInstanceId
                     FROM ProcessInstanceLog spl
                     WHERE l.processinstanceid = spl.processinstanceid);
```

produces

```
Delete on audittaskimpl l  (cost=8.61..424652.49 rows=38144 width=6)
  ->  Seq Scan on audittaskimpl l  (cost=8.61..424652.49 rows=38144 width=6)
        Filter: ((hashed SubPlan 1) OR (NOT (SubPlan 2)))
        SubPlan 1
          ->  Index Scan using idx_pinstlog_status on processinstancelog spl  (cost=0.29..8.61 rows=1 width=8)
                Index Cond: (status = ANY ('{2,3}'::integer[]))
        SubPlan 2
          ->  Index Only Scan using idx_pinstlog_pinstid on processinstancelog spl_1  (cost=0.29..8.31 rows=1 width=0)
                Index Cond: (processinstanceid = l.processinstanceid)

```

Original SQL query:
```
EXPLAIN DELETE
FROM AuditTaskImpl l
WHERE l.processInstanceId IN (SELECT spl.processInstanceId
                              FROM ProcessInstanceLog spl
                              WHERE spl.status in (2,3))
```

produces

```
Delete on audittaskimpl l  (cost=8.62..2147.72 rows=1 width=12)
  ->  Hash Semi Join  (cost=8.62..2147.72 rows=1 width=12)
        Hash Cond: (l.processinstanceid = spl.processinstanceid)
        ->  Seq Scan on audittaskimpl l  (cost=0.00..2005.59 rows=50859 width=14)
        ->  Hash  (cost=8.61..8.61 rows=1 width=14)
              ->  Index Scan using idx_pinstlog_status on processinstancelog spl  (cost=0.29..8.61 rows=1 width=14)
                    Index Cond: (status = ANY ('{2,3}'::integer[]))
```

So cost is apparently much much lower. But anyway, it was a good exercise. At least I have found ignored tests which work so will alter the PR to just unignore them, wdyt?

Reply:
Ok, my last call, I have managed to optimize it a little bit :)
```
EXPLAIN DELETE
FROM AuditTaskImpl l
WHERE NOT EXISTS (SELECT spl.processInstanceId
                              FROM ProcessInstanceLog spl
                              WHERE l.processinstanceid = spl.processinstanceid
                              AND spl.status NOT IN (2,3))
```

it now gives:

```
Delete on audittaskimpl l  (cost=2793.65..5864.11 rows=1 width=12)
  ->  Hash Anti Join  (cost=2793.65..5864.11 rows=1 width=12)
        Hash Cond: (l.processinstanceid = spl.processinstanceid)
        ->  Seq Scan on audittaskimpl l  (cost=0.00..2005.59 rows=50859 width=14)
        ->  Hash  (cost=1908.83..1908.83 rows=50866 width=14)
              ->  Seq Scan on processinstancelog spl  (cost=0.00..1908.83 rows=50866 width=14)
                    Filter: (status <> ALL ('{2,3}'::integer[]))

```

I think the original one might be more costly if there are more completed processes than active ones, mine will be more costly if there are more active processes I guess.